ENH: Add SIMD versions of bool logical_&&,||,! and absolute

Developer-Ecosystem-Engineering opened the pull request:
NumPy has SIMD versions of BOOL `logical_and`, `logical_or`, `logical_not`, and `absolute` for SSE2.  The changes here replace that implementation with one that uses universal intrinsics.  This allows other architectures to have SIMD versions of the functions too.

BOOL `logical_and` and `logical_or` are particularly important for NumPy as that's how  `np.any()` / `np.all()` are implemented.


Apple M1: **up to 16.5x faster**
```
       before           after         ratio
     [7c143834]       [c49d9dc2]
     <main>           <logical/dev>
-        3.47±0μs      1.82±0.01μs     0.52  bench_reduce.AnyAll.time_all_slow
-     4.05±0.01μs      1.83±0.06μs     0.45  bench_reduce.AnyAll.time_any_slow
-        6.55±0μs          532±2ns     0.08  bench_ufunc.Custom.time_not_bool
-        10.2±0μs          680±7ns     0.07  bench_ufunc.Custom.time_or_bool
-     11.0±0.07μs          665±3ns     0.06  bench_ufunc.Custom.time_and_bool
                                                                                                                                                                    
SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

Apple M1 (Rosetta): **up to 1.6x faster**
```
       before           after         ratio
     [7c143834]       [1ad2f2f2]
     <main>           <logical/dev>
-     1.14±0.01μs         1.03±0μs     0.90  bench_ufunc.Custom.time_not_bool
-     4.38±0.03μs      3.06±0.02μs     0.70  bench_reduce.AnyAll.time_any_slow
-     5.20±0.01μs      3.17±0.01μs     0.61  bench_reduce.AnyAll.time_all_slow

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

iMac Pro (AVX512): **up to 1.2x faster**
```
       before           after         ratio
     [da6297b9]       [c49d9dc2]
     <main>           <logical/dev>
-     1.42±0.02μs      1.24±0.03μs     0.87  bench_ufunc.Custom.time_and_bool
-     4.16±0.03μs       3.60±0.1μs     0.86  bench_reduce.AnyAll.time_any_slow
-     1.21±0.03μs      1.03±0.02μs     0.86  bench_ufunc.Custom.time_not_bool
-      4.30±0.1μs      3.56±0.07μs     0.83  bench_reduce.AnyAll.time_all_slow

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

Developer-Ecosystem-Engineering committed:
ENH: Add SIMD versions of bool logical_and, logical_or, logical_not and absolute

NumPy has SIMD versions of BOOL `logical_and`, `logical_or`, `logical_not`, and `absolute` for SSE2.  The changes here replace that implementation with one that uses their universal intrinsics.  This allows other architectures to have SIMD versions of the functions too.

BOOL `logical_and` and `logical_or` are particularly important for NumPy as that's how  `np.any()` / `np.all()` are implemented.

mattip commented on numpy/core/setup.py.orig:
```diff
@@ -0,0 +1,1173 @@
+import os
```
This backup of `setup.py` should be removed.

seiko2plus replied:
still

mattip reviewed (commented):
Nice. I will defer to @seiko2plus for the deeper SIMD review. A couple of general questions for all the latest SIMD PRs:
- could you check the binary file size change of `_multiarray_umath*.so` after vs. before?
- Did the change have an effect on any other benchmarks besides the obvious ones that you reported
- If possible, could you run the benchmarks on a non-avx512 x86_64 system

seiko2plus commented on numpy/core/src/umath/loops_logical.dispatch.c.src:
```diff
@@ -0,0 +1,427 @@
+ ** Extra SIMD intrinsics
+ ******************************************************************************/
+
+#if NPY_SIMD
+#if !defined(NPY_HAVE_SSE2)
+#define USE_NPYV_REDUCE_MINMAX
+#endif
+
+#if defined(NPY_HAVE_ASIMD) && defined(__aarch64__)
+    #define npyv_reduce_min_u8 vminvq_u8
+    #define npyv_reduce_max_u8 vmaxvq_u8
+#elif defined(USE_NPYV_REDUCE_MINMAX)
+    // Scalar intrinsics
+    #define scalar_max_i(A, B) ((A > B) ? A : B)
+    #define scalar_min_i(A, B) ((A < B) ? A : B)
+
+    /**begin repeat
+     * #intrin = min, max#
+     */
+    NPY_FINLINE npyv_lanetype_u8 npyv_reduce_@intrin@_u8(npyv_u8 v)
+    {
+        npyv_lanetype_u8 NPY_DECL_ALIGNED(NPY_SIMD_WIDTH) s[npyv_nlanes_u8];
+        npyv_storea_u8(s, v);
+        npyv_lanetype_u8 result = s[0];
+        for(int i=1; i<npyv_nlanes_u8; ++i){
+            result = scalar_@intrin@_i(result, s[i]);
+        }
+        return result;
+    }
+    /**end repeat**/
+    #undef scalar_max_i
+    #undef scalar_min_i
+#endif
+#endif // NPY_SIMD
```
```suggestion
```
covered by #22306, in case if still required.

seiko2plus commented on numpy/core/src/umath/loops_logical.dispatch.c.src:
```diff
@@ -0,0 +1,427 @@
+ ** Extra SIMD intrinsics
+ ******************************************************************************/
+
+#if NPY_SIMD
+#if !defined(NPY_HAVE_SSE2)
+#define USE_NPYV_REDUCE_MINMAX
+#endif
```
```suggestion
```
no need for two paths, new intrinsics any/all covered by #22306 should provide fair optimization across all SIMD extensions

seiko2plus reviewed (changes requested):
Sorry for the delayed response. For fair optimization and to avoid two separate implementations for reduction, new universal intrinsic any/all going to be implemented through pr #22306.

Developer-Ecosystem-Engineering committed:
Update numpy/core/src/umath/loops_logical.dispatch.c.src

Co-authored-by: Sayed Adel <seiko@imavr.com>

Developer-Ecosystem-Engineering committed:
Update numpy/core/src/umath/loops_logical.dispatch.c.src

Co-authored-by: Sayed Adel <seiko@imavr.com>

Developer-Ecosystem-Engineering committed:
Update numpy/core/src/umath/loops_logical.dispatch.c.src

Co-authored-by: Sayed Adel <seiko@imavr.com>

Developer-Ecosystem-Engineering commented:
I merged in the suggestions made to adopt changes in #22306 (didn't exist at the time of this PR authorship)

Developer-Ecosystem-Engineering committed:
Add loops_logical.dispatch.c.src to meson.build

Developer-Ecosystem-Engineering committed:
Finish adopting universal intrinsics to fix AVX2/AVX512 tests.

Developer-Ecosystem-Engineering committed:
Fix smoke test running without NPY_SIMD

seiko2plus commented on numpy/core/src/umath/loops_logical.dispatch.c.src:
```diff
@@ -0,0 +1,349 @@
+    // Single vectors loop
+    for (; len >= vstep; len -= vstep, ip += vstep) {
+        npyv_u8 v0 = npyv_load_u8(ip);
+        npy_uint8 r = npyv_reduce_@reduce@_u8(v0);
+        if(r @scalar_cmp@ 0){
```
```suggestion
        if ([! , ] npyv_[all, any]_u8(v0)) {
```
should provide better performance across all SIMD extensions

seiko2plus commented on numpy/core/src/umath/loops_logical.dispatch.c.src:
```diff
@@ -0,0 +1,349 @@
+}
+/**end repeat**/
+
+#undef npyv_reduce_min_u8
+#undef npyv_reduce_max_u8
```
```suggestion
```

seiko2plus reviewed (changes requested):
Excellent work, just one nit more.

Developer-Ecosystem-Engineering committed:
Remote setup.py.orig and use npyv_any/all

seiko2plus reviewed (changes requested):
My apologies, It seems I wasn't entirely focused. One thing more.

Developer-Ecosystem-Engineering committed:
Use mask_to_true() for logical_not to save a few ops

seiko2plus committed:
ENG, SIMD: Dispatch bool (invert, add, fmax, fmin...)

  Following functions are defined by umath generator
  to enable runtime dispatching without the need
  to redefine them within dsipatch-able sources:

    BOOL_invert, BOOL_add, BOOL_bitwise_and
    BOOL_bitwise_or, BOOL_logical_xor
    BOOL_bitwise_xor, BOOL_multiply
    BOOL_maximum, BOOL_minimum, BOOL_fmax,
    BOOL_fmin

seiko2plus reviewed (approved):
LGTM, Thank you.
I have made some additional changes to umath generator to enable runtime dispatching for the following alias:
```
BOOL_invert, BOOL_add, BOOL_bitwise_and
BOOL_bitwise_or, BOOL_logical_xor
BOOL_bitwise_xor, BOOL_multiply
BOOL_maximum, BOOL_minimum, BOOL_fmax,
BOOL_fmin
```

seiko2plus commented:
The last push after the approval to satisfy python linter and to fix dispatch karg on umath generator for `logical_xor`, both errors were caused by my latest patch.

Developer-Ecosystem-Engineering commented:
@mattip Anything else missing/required?

mattip commented:
Thanks @Developer-Ecosystem-Engineering and @seiko2plus

tylerjereddy commented:
Note that this showed up as a bit naughty in `git bisect` from gh-22845.
